bin/client/src/cached_eigenda_provider.rs

```diff
@@ -0,0 +1,136 @@
+use kona_preimage::CommsClient;
+use kona_preimage::errors::PreimageOracleError;
+use alloy_primitives::Bytes;
+use async_trait::async_trait;
+use alloy_rlp::Decodable;
+
+use kona_proof::errors::OracleProviderError;
+use hokulea_proof::eigenda_provider::OracleEigenDAProvider;
+use hokulea_eigenda::EigenDABlobProvider;
+use hokulea_eigenda::BlobInfo;
+
+use crate::witness::EigenDABlobWitness;
+
+use rust_kzg_bn254::kzg::KZG;
+use rust_kzg_bn254::blob::Blob;
+use num::BigUint;
+
+
+#[derive(Debug, Clone)]
+pub struct CachedOracleEigenDAProvider<T: CommsClient> {
+    /// The preimage oracle client.
+    oracle: OracleEigenDAProvider<T>,
+    /// kzg proof witness
+    witness: EigenDABlobWitness,
+}
+
+impl<T: CommsClient> CachedOracleEigenDAProvider<T> {
+    /// Constructs a new oracle-backed EigenDA provider.
+    pub fn new(oracle: OracleEigenDAProvider<T>, witness: EigenDABlobWitness) -> Self {
+        Self { oracle, witness }
+    }
+}
+
+#[async_trait]
+impl <T: CommsClient + Sync + Send> EigenDABlobProvider for CachedOracleEigenDAProvider<T> {
+    type Error = OracleProviderError;
+
+    async fn get_blob(&mut self, cert: &Bytes) -> Result<Bytes, Self::Error> {
+        match self.oracle.get_blob(cert).await {
+            Ok(b) => {
+                let item_slice = cert.as_ref();
+                let cert_blob_info = match BlobInfo::decode(&mut &item_slice[4..]) {
+                    Ok(c) => c,
+                    Err(_) => return Err(OracleProviderError::Preimage(
+                        PreimageOracleError::Other("does not contain header".into(),
+                    ))),
+                };
+
+                let output = self.compute_and_save_witness(&b)?;
```

Comment:
don't use generic names like "output". better to use self-documenting names, makes the code much more easily readable.